### DISCUSSION TITLE
Configurable rectangle positions

### DISCUSSION THREAD

**Discussion by codycraven** [rxhanson/Rectangle#427] [Ideas]
First thanks for the application! I noticed many, many of the feature requests are items to support additional placement locations or widths (9ths, 6ths, etc).

I have a large display and would like to set up custom placements, something that would look like this (with each row being 1/2 screen height):
```
|    1/2    |    1/2    |
| 1/8 | 3/8 | 3/8 | 1/8 |
```

I could achieve this by forking the project and defining my own custom rectangle positions/hotkeys, however it seems like this would be great to be exposed through configuration. This would hopefully stop all of the requests for additional rectangle sizes/positions since users could define their own.

This could be configured through a JSON format along these lines to define a layout (or any other config file format desired):
```
{
  layouts:  {
    default: {
      // Layout label would show in Menu/Hotkey UI to switch between layout
      // configurations
      label: "Default",
      // Rectangles would define the rectanges for the layout.
      // This property could be an array instead of an object, I chose an object
      // for the example to give future flexibility, if you wanted to support
      // moving windows to specific rectangles automatically or something.
      rectangles: {
        topLeftSixth: {
          label: "Top Left Sixth",
          // I defined all positions as fractional values for programming
          // simplicity, could instead be a string if you wanted to support
          // different formats, such as "1/2", "128px", "50%", etc.
          top: 0,
          bottom: .5,
          left: 0,
          right: .834
          // Could record hotkey on rectangle definitions if desired.
          // Would allow user to port configuration from one system to another
          // without having to manually reset hotkeys.
          hotkey: "⌘⌥⇧u",
        },
        // All other currently supported rectangles would be defined here.
      },
      // Again if you want to support hotkeys in the config you could set a
      // hotkey to switch to the layout.
      hotkey: "⌘⌥⇧1",
    },
    // Here would be a user's custom layout.
    myCustomLayout: {
      label: "My Custom Layout",
      rectangles: {
        topLeft: {
          // Position label would be value shown in Hotkey/Menu UI
          label: "Top Left",
          top: 0,
          bottom: 0.5,
          left: 0,
          right: 0.5,
          // Hotkey used in another layout could be re-used here without
          // conflict because it would be in the context of the current layout.
          hotkey: "⌘⌥⇧u",
        },
        topRight: {
          label: "Top Right",
          top: 0,
          bottom: 0.5,
          left: 0.5,
          hotkey: "⌘⌥⇧p",
        },
        bottomLeft: {
          label: "Bottom Left Eighth",
          top: 0.5,
          bottom: 0,
          left: 0,
          right: 0.875,
          hotkey: "⌘⌥⇧j",
        },
        indentedBottomLeft: {
          label: "Indented Bottom Left Three-Eigths",
          top: 0.5,
          bottom: 0,
          left: 0.125,
          right: 0.5,
          hotkey: "⌘⌥⇧k",
        },
        indentedBottomRight: {
          label: "Indented Bottom Right Three-Eights",
          top: 0.5,
          bottom: 0,
          left: 0.5,
          right: 0.125,
          hotkey: "⌘⌥⇧l",
        },
        bottomRight: {
          label: "Bottom Right",
          top: 0.5,
          bottom: 0,
          left: 0.875,
          right: 0,
          hotkey: "⌘⌥⇧;",
        },
      },
      hotkey: "⌘⌥⇧2",
    }
    }
  },
  movement: {
    // Movement related actions could be defined here, such as "Move Up",
    // "Move Left", "Maximize", "Next Display", etc.
    //
    // Although they don't have to be, I wanted to show a concept this format
    // would support for future features.
  }
}
```

**Comment by rxhanson** [comment]
Thanks for the suggestion! Since config file based window management is achieved by Hammerspoon, I haven't felt a strong urge to add this in to Rectangle. My thought is that most people that want this type of thing would eventually just gravitate to Hammerspoon since there's a decently large user base and probably tons of existing config files floating around that you could grab. 

As you might expect provided the quantity of requests for additional window sizes/positions, this is something that I have given a lot of thought. My implementation of this is in a custom shortcut builder in my paid superset of Rectangle, Hookshot (https://hookshot.app). It contains the ability to declare what window size/position you want on repeated executions of the shortcuts as well. Here's a post I wrote about the custom shortcut builder, how to use it, and why it's not in Rectangle: https://medium.com/ryan-hanson/custom-window-dimension-shortcuts-in-macos-e7df347dfd24. 

With all of that said, if anyone wants to make a pull request to allow additional JSON config for additional window size/position shortcuts, I don't see why I would turn that down.

**Reply by codycraven** [reply]
@rxhanson thanks for the context. I'm not generally a Mac user and wasn't aware that Hammerspoon could be used for window management. I was solely coming from Yabai/Amethyst (neither of which let me have any sort of defined layout) and Rectangles filled my needs as a much more suitable option.

As a note, I did take a look at Hookshot yesterday, my take away was it was a tool for mouse/touch gesture support (something I don't personally make much use of) and as such I dismissed it without looking further and realizing that you have this sort of functionality available in it.

I definitely would not want to encroach on the positioning of your paid product by pressing for Rectangle to be something you didn't design it to be.

Thanks again!

**Reply by rxhanson** [reply]
Thanks for the follow up! 

Just to be clear on where I see Hookshot in relation to Rectangle: It is not my intent to keep features out of Rectangle so that users will be inclined to purchase Hookshot. To me, Hookshot is basically a place where I can try out more advanced features that I really like and use, and Rectangle is purposefully kept simple (the everyman's window manager). That way, Rectangle is approachable, not overwhelming with features, and simple for people to use the code - either by contributing or by just using it in their own projects. The custom builder added a lot more complexity than I had initially expected, and that's why it's not in Rectangle. 

Since Hammerspoon and Hookshot are existing alternative solutions, I don't feel compelled to add this in myself, but again - I don't see a reason to turn down a PR for it since it wouldn't add unnecessary complexity to the app for an end user.
